File: lipnet-beam.py
'''

LipNet: End-to-End Sentence-level Lipreading. arXiv preprint arXiv:1611.01599 (2016).

Copyright (C) 2017 Yannis M. Assael, Brendan Shillingford, Shimon Whiteson, Nando de Freitas

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

'''

import collections
import h5py
import multiprocessing as mp
import numpy as np
import os
import pickle
import sys

os.environ['OMP_NUM_THREADS'] = '1'

vocab = '- abcdefghijklmnopqrstuvwxyz'


def int_to_char(int_seq):
    return ''.join([vocab[i] for i in int_seq])


def ngram(b, n):
    return set([b[i:i + n] for i in range(len(b) - n + 1)])


def ngram_full(b, n_start, n_stop):
    ret = set()
    for n in range(n_start, n_stop + 1):
        ret = ret | set([b[i:i + n] for i in range(len(b) - n + 1)])
    return ret


class Hyp(object):
    """
    Container class for a single hypothesis
    """

    def __init__(self, pb, pnb, nc):
        self.p_b = pb
        self.p_nb = pnb
        self.n_c = nc


class DecoderBase(object):
    """
    This is the base class for any decoder.
    It does not perform actual decoding (think of it as an abstract class)
    To make your decoder work, extend the class and implement the decode function
    """

    # __metaclass__ = abc.ABCMeta

    # TODO are static methods mixed with cython still fast?
    # @staticmethod
    def combine(self, a, b, c=float('-inf')):
        psum = np.exp(a) + np.exp(b) + np.exp(c)
        if psum == 0.0:
            return float('-inf')
        else:
            return np.log(psum)

    # character mapping objects. Cython requires declaring in advance
    def load_chars(self, charmap_file):
        """
        Loads a mapping of character -> int
        Stores mapping in self.char_int_map
        Stores int -> char mapping in self.int_char_map
        returns True if maps created successfully
        """
        with open(charmap_file) as fid:
            self.char_int_map = dict(
                tuple(l.strip().split()) for l in fid.readlines())

        self.int_char_map = {}
        for k, v in self.char_int_map.items():
            self.char_int_map[k] = int(v)
            self.int_char_map[int(v)] = k

        return True

    # cpdef seq_int_to_char(self, )
    # @abc.abstractmethod
    def decode(self, probs):
        """
        Child classes must implement the decode function
        Minimally the decode function takes a matrix of probabilities
        output by the network (characters vs time)
        returns the best hypothesis in characters
        """
        return None


class ArgmaxDecoder(DecoderBase):
    """
    This is the simple argmax decoder. It doesn't need an LM
    It performs basic collapsing decoding
    """

    def decode(self, probs):
        """
        Takes matrix of probabilities and computes per-frame argmax
        Applies basic blank/duplicate collapsing
        returns the best hypothesis in characters
        Charmap must be loaded 
        """
        maxInd = np.argmax(probs, axis=0)
        pmInd = -1
        hyp = []
        # TODO is this the right way to score argmax decoding?
        hyp_score = 0.0
        for t in range(probs.shape[1]):
            hyp_score = hyp_score + probs[maxInd[t], t]
            if maxInd[t] != pmInd:
                pmInd = maxInd[t]
                if pmInd > 0:
                    hyp.append(self.int_char_map[pmInd])

        # collapsed hypothesis (this is our best guess)
        hyp = ''.join(hyp)
        return hyp, hyp_score


class BeamLMDecoder(DecoderBase):
    """
    Beam-search decoder with character LM
    """

    def load_lm(self, lmfile):
        """
        Loads a language model from lmfile
        returns True if lm loading successful
        """
        self.lm = pickle.load(open("lm/lm_2gram.p", "rb")) | pickle.load(open("lm/lm_3gram.p", "rb")
                                                                         ) | pickle.load(
            open("lm/lm_4gram.p", "rb")) | pickle.load(open("lm/lm_5gram.p", "rb"))
        return True

    def lm_score_final_char(self, prefix, query_char, alpha):
        """
        uses lm to score entire prefix
        returns only the log prob of final char
        """
        if alpha > 0:
            # convert prefix and query to actual text
            full_int = list(prefix) + [query_char]

            full_str = ''.join([self.int_char_map[i] for i in full_int])

            if full_str[0] not in ['b', 'p', 'l', 's']:
                return float('-inf')

            # print(full_str)
            for g in ngram_full(full_str, 2, 5):
                if g not in self.lm:
                    return float('-inf')

        return 0

    def decode(self, probs, beam=40, alpha=1.0, beta=0.0):
        """
        Decoder with an LM
        returns the best hypothesis in characters
        Charmap must be loaded 
        """
        N = probs.shape[0]
        T = probs.shape[1]

        keyFn = lambda x: self.combine(x[1][0], x[1][1]) + beta * x[1][2]
        initFn = lambda: [float('-inf'), float('-inf'), 0]

        # [prefix, [p_nb, p_b, node, |W|]]
        Hcurr = [[(), [float('-inf'), 0.0, 0]]]
        Hold = collections.defaultdict(initFn)

        # loop over time
        for t in range(T):
            Hcurr = dict(Hcurr)
            Hnext = collections.defaultdict(initFn)

            for prefix, (v0, v1, numC) in Hcurr.items():

                valsP = Hnext[prefix]
                valsP[1] = self.combine(
                    v0 + probs[0, t], v1 + probs[0, t], valsP[1])
                valsP[2] = numC
                if len(prefix) > 0:
                    valsP[0] = self.combine(v0 + probs[prefix[-1], t], valsP[0])

                for i in range(1, N):
                    nprefix = tuple(list(prefix) + [i])
                    valsN = Hnext[nprefix]

                    # query the LM_SCORE_FINAL_CHAR for final char score
                    lm_prob = self.lm_score_final_char(prefix, i, alpha)
                    # lm_prob = alpha*lm_placeholder(i,prefix)
                    valsN[2] = numC + 1
                    if len(prefix) == 0 or (len(prefix) > 0 and i != prefix[-1]):
                        valsN[0] = self.combine(
                            v0 + probs[i, t] + lm_prob, v1 + probs[i, t] + lm_prob, valsN[0])
                    else:
                        valsN[0] = self.combine(
                            v1 + probs[i, t] + lm_prob, valsN[0])

                    if nprefix not in Hcurr:
                        v2, v3, _ = Hold[nprefix]
                        valsN[1] = self.combine(
                            v2 + probs[0, t], v3 + probs[0, t], valsN[1])
                        valsN[0] = self.combine(v2 + probs[i, t], valsN[0])

            Hold = Hnext
            Hcurr = sorted(Hnext.items(), key=keyFn, reverse=True)[:beam]

        hyp = ''.join([self.int_char_map[i] for i in Hcurr[0][0]])

        return hyp  # , keyFn(Hcurr[0])
        # return hyp
        # return list(Hcurr[0][0]),keyFn(Hcurr[0])


if __name__ == '__main__':
    if len(sys.argv) > 1:
        beam_width = int(sys.argv[1])
    else:
        beam_width = 100

    if len(sys.argv) > 2:
        filename = sys.argv[2]
    else:
        filename = '/tmp/pctc.h5'

    if len(sys.argv) > 3:
        alpha = float(sys.argv[3])
    else:
        alpha = 1

    if len(sys.argv) > 4:
        beta = float(sys.argv[4])
    else:
        beta = 0

    # filename = 'pctc.h5'

    # Load probs
    pctc = h5py.File(filename, 'r')['/pctc'][()]

    ctc_beam = BeamLMDecoder()

    ctc_beam.int_char_map = {}
    ctc_beam.char_int_map = {}
    for i in range(len(vocab)):
        ctc_beam.int_char_map[i] = vocab[i]
        ctc_beam.char_int_map[vocab[i]] = i

    ctc_beam.load_lm('lm/lm3.klm')


    def ctc_bs(x):
        return ctc_beam.decode(x.T, beam=beam_width, alpha=alpha, beta=beta)


    ctc_bs_in = []
    for b in range(pctc.shape[1]):
        pctc_log = np.log(pctc[:, b, :].squeeze())
        ctc_bs_in.append(pctc_log)

    # print(ctc_bs(ctc_bs_in[0]))

    pool = mp.Pool(processes=100)
    results = []
    imap_res = pool.imap(ctc_bs, ctc_bs_in)
    for r in imap_res:
        results.append(r)

    pool.close()
    pool.join()

    for r in results:
        print(r)

File: lipnet-beam.lua
--[[

LipNet: End-to-End Sentence-level Lipreading. arXiv preprint arXiv:1611.01599 (2016).

Copyright (C) 2017 Yannis M. Assael, Brendan Shillingford, Shimon Whiteson, Nando de Freitas

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

]] --

--
-- Dependencies
--

require 'io'
require 'sys'
require 'nn'
require 'nngraph'
require 'optim'
require 'hdf5'
require 'paths'

require 'cutorch'
require 'cunn'
require 'cunnx'
require 'cudnn'
cudnn.benchmark = true
cudnn.fastest = true

require 'modules.TemporalJitter'

require 'warp_ctc'
assert(gpu_ctc, 'requires Baidu unmodified warp_ctc (not my fork), built with GPU support')

require 'nnx'

local log = require 'util.log'
log.level = "debug"

require 'pprint'


--
-- Configuration
--

local cmd = torch.CmdLine()
cmd:text()
cmd:text('Options')
cmd:option('-beam_width', 200, '')
cmd:option('-seed', 123, 'initial random seed')
cmd:option('-debug', 0, '')
cmd:option('-verbose', 0, '')
cmd:option('-beam_alpha', 1, '')
cmd:option('-beam_beta', 0, '')
cmd:option('-datapath', 'data', 'video data path')
cmd:option('-bs', 100, 'batch size')
cmd:option('-checkpoint', '', '')
cmd:option('-rnn_size', 256, 'rnn size')
cmd:option('-dropout', 0.5, '')
cmd:option('-threads', 8, 'number of torch built-in threads')
cmd:option('-exp', 'exp.0001', 'lua file that returns model, experiment name, optimization settings, ...')
cmd:option('-exp_i', 1, 'if exp returns more than one model, uses this index')
cmd:option('-ignore_checkpoint', 1, 'do not continue from checkpoint')
cmd:option('-print_every', 1, 'iterations between printing')
cmd:option('-test_every', 1, 'iterations between testing')
cmd:option('-checkpoint_every', 1, 'iterations between saving checkpoints')
cmd:text()
local opt = cmd:parse(arg)
assert(opt.exp ~= '', 'exp lua file required')
for k, v in pairs(opt) do
    log.infof('opt: %s=%s', k, tostring(v))
end

--
-- Initialisation
--
torch.manualSeed(opt.seed)
torch.setnumthreads(opt.threads)
torch.setdefaulttensortype('torch.FloatTensor')

--
-- CUDA
--

torch.manualSeed(opt.seed)
cutorch.manualSeed(opt.seed)

--
-- Load model
--
local checkpoint = torch.load(opt.checkpoint)

local model = checkpoint.model
local opt_ch = checkpoint.opt
opt_ch.debug = 0
opt_ch.bs = opt.bs
local stats = checkpoint.stats
local iter = checkpoint.iter

pprint(opt_ch)

-- Load exp
local exp = require(opt_ch.exp)(opt_ch)
pprint(exp.data_loader.vocab)

function editDistance(prediction, target)
    local d = torch.Tensor(#target + 1, #prediction + 1):zero()
    for i = 1, #target + 1 do
        for j = 1, #prediction + 1 do
            if (i == 1) then
                d[1][j] = j - 1
            elseif (j == 1) then
                d[i][1] = i - 1
            end
        end
    end

    for i = 2, #target + 1 do
        for j = 2, #prediction + 1 do
            if (target[i - 1] == prediction[j - 1]) then
                d[i][j] = d[i - 1][j - 1]
            else
                local substitution = d[i - 1][j - 1] + 1
                local insertion = d[i][j - 1] + 1
                local deletion = d[i - 1][j] + 1
                d[i][j] = torch.min(torch.Tensor({ substitution, insertion, deletion }))
            end
        end
    end
    local errorRate = d[#target + 1][#prediction + 1] / #target
    return errorRate
end

-- load data
local softmax = nnob.TimeBatchWrapper { mod = nn.SoftMax() }:cuda()

local stats = { beam_cer = 0, beam_wer = 0 }

-- Test
model.pred:evaluate()

local bs_count = 0

for it = 1, #exp.data_loader.dataset_val, opt.bs do

    xlua.progress(it, #exp.data_loader.dataset_val)

    -- batch indexes
    local idx = {}
    for i = 0, opt.bs - 1 do
        if it + i <= #exp.data_loader.dataset_val then
            table.insert(idx, it + i)
        end
    end

    -- load data
    local x, y, lengths = unpack(exp.data_loader:forward(idx, true))

    -- predict data
    local logits = model.pred:forward(x)
    local prob = softmax:forward(logits):float()

    -- save data
    local filename = '/tmp/pctc.h5'
    local pctcFile = hdf5.open(filename, 'w')
    pctcFile:write('/pctc', prob)
    pctcFile:close()

    -- Beam Search
    local handle = io.popen('python lipnet-beam.py ' .. opt.beam_width .. ' ' .. filename .. ' ' .. opt.beam_alpha .. ' ' .. opt.beam_beta)
    local beam = handle:read("*a"):split("\n")
    handle:close()

    -- Compute error
    for b = 1, #idx do
        bs_count = bs_count + 1

        -- Target
        local tok_y = {}
        for t = 1, #y[b] do
            table.insert(tok_y, exp.data_loader.vocab[y[b][t]])
        end

        if opt.verbose and it == 1 then
            print(b, 'y', table.concat(tok_y, ""))
        end

        -- Split in chars
        local tok = {}
        for c in beam[b]:gmatch "." do
            table.insert(tok, c)
        end


        -- Compute WER
        local tok_w = table.concat(tok, ""):split(' ')
        local tok_y_w = table.concat(tok_y, ""):split(' ')
        local wer = editDistance(tok_w, tok_y_w)
        stats.beam_wer = stats.beam_wer + wer

        -- Compute CER
        local cer = editDistance(tok, tok_y)
        stats.beam_cer = stats.beam_cer + cer
        if opt.verbose and it == 1 then
            print(b, 'b' .. opt.beam_width, table.concat(tok, ""), cer, wer)
        end

        -- Output
        local tok = {}
        local p = prob:narrow(2, b, 1):squeeze()
        for t = 1, lengths[b] do
            local _, c = torch.max(p[t], 1)
            c = c:squeeze()
            table.insert(tok, exp.data_loader.vocab[c - 1] or '_')
        end
        if it == 1 then
            print(b, 'o', table.concat(tok, ""))

            if opt.verbose then
                print('')
            end
        end

        -- break
    end
end


print(opt.beam_width, opt.beam_alpha, opt.beam_beta, 'beam_width CER', stats.beam_cer / bs_count)
print(opt.beam_width, opt.beam_alpha, opt.beam_beta, 'beam_width WER', stats.beam_wer / bs_count)
